### PR TITLE
Routing system

### PR DESCRIPTION
**Pull request by Naomi-Dennis**
The routing system is being changed from an if/else system to an object action system. 
Additions:
* Method verbs are considered when processing requests
* Actions are created via `Application` objects and instantiated in the `Main` class
* The `Router` is the jump off point for applications. It is not longer `SimpleApplication`

### CODE REVIEW

File: src/test/java/http/RouterIdTest.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package http;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class RouterIdTest {
```
Comment (damonkelley) [inline src/test/java/http/RouterIdTest.java]: `RouteIdTest`?

File: src/main/java/http/Router.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    public HttpResponse start(HttpRequest clientRequest){
+        RouteId id = new RouteId(clientRequest.getMethod(), clientRequest.getRequestPath());
+
+        Optional<Application> requestedApplication = Optional.ofNullable(routes.get(id));
+        Application verifiedApplication = requestedApplication.orElse(new NotFound());
+        HttpResponse applicationResponse = verifiedApplication.start(clientRequest);
+
+        return applicationResponse;
```
Comment (damonkelley) [inline src/main/java/http/Router.java]: *This is purely aesthetics*

I really like the way it reads when optionals are chained. Something like

```java
public HtttpResponse start(HttpRequest clientRequest) {
    RouteId id = new RouteId(clientRequest.getMethod(), clientRequest.getRequestPath());

    return Optional.ofNullable(routes.get(id))
        .orElse(new NotFound())
        .start(clientRequest);
}
```

It reads top to bottom in a way that is easy to understand. Bonus is that you don't have come up with a variable name for each intermediate stage (i.e. `verifiedApplication`).

File: src/main/java/main/Main.java [hunk 3]
```diff
@@ -26,7 +30,14 @@ public static void main(String[] args) throws IOException {
...
 
-        Http httpProcess = new Http(new SimpleApplication(), logger);
+        Map<RouteId, Application> routes = Map.ofEntries(
+                Map.entry(new RouteId(HttpMethod.GET, "/not_found"), new NotFound()),
+                Map.entry(new RouteId(HttpMethod.GET, "/simple_get"), new SimpleGet()),
+                Map.entry(new RouteId(HttpMethod.POST, "/echo_body"), new EchoBody()),
+                Map.entry(new RouteId(HttpMethod.GET, "/redirect"), new Redirect())
+        );
+
+        Http httpProcess = new Http(new Router(routes), logger);
```
Comment (damonkelley) [inline src/main/java/main/Main.java]: Nice!

File: src/test/java/http/RouterIdTest.java (hunk 1 above)
Comment (Naomi-Dennis) [reply src/test/java/http/RouterIdTest.java]: It's the test for `RouterId`

**Pull request merged**